cirq/ion/ion_device.py

```diff
@@ -0,0 +1,211 @@
+                raise ValueError('Unsupported qubit type: {!r}'.format(q))
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+    def _check_if_XXPow_operation_interacts_with_any(
+            self,
+            XXPow_op: ops.GateOperation,
+            others: Iterable[ops.GateOperation]) -> bool:
+        return any(self._check_if_XXPow_operation_interacts(XXPow_op, op)
+                   for op in others)
+
+    def _check_if_XXPow_operation_interacts(
+            self,
+            XXPow_op: ops.GateOperation,
+            other_op: ops.GateOperation) -> bool:
+        if isinstance(other_op.gate, (ops.XPowGate,
+                                      ops.YPowGate,
+                                      ops.PhasedXPowGate,
+                                      ops.MeasurementGate,
+                                      ops.ZPowGate)):
+            return False
+
+        return any(q == p
+                   for q in XXPow_op.qubits
+                   for p in other_op.qubits)
+
+    def validate_scheduled_operation(self, schedule, scheduled_operation):
+        self.validate_operation(scheduled_operation.operation)
+
+        if isinstance(scheduled_operation.operation.gate, ops.XXPowGate):
+            for other in schedule.operations_happening_at_same_time_as(
+                    scheduled_operation):
+                if self._check_if_XXPow_operation_interacts(
+                        cast(ops.GateOperation, scheduled_operation.operation),
+                        cast(ops.GateOperation, other.operation)):
+                    raise ValueError(
+                        'Simultaneous two-qubit '
+                        'operations on same qubit: {} vs {}.'.format(
+                            scheduled_operation, other))
+
+    def validate_circuit(self, circuit: circuits.Circuit):
+        super().validate_circuit(circuit)
+        _verify_unique_measurement_keys(circuit.all_operations())
+
+    def can_add_operation_into_moment(self,
+                                      operation: ops.Operation,
+                                      moment: ops.Moment) -> bool:
+
+        if not super().can_add_operation_into_moment(operation, moment):
+            return False
+        if (isinstance(operation, ops.GateOperation) and
+                isinstance(operation.gate, ops.XXPowGate)):
+            return not self._check_if_XXPow_operation_interacts_with_any(
+                cast(ops.GateOperation, operation),
+                cast(Iterable[ops.GateOperation], moment.operations))
+        return True
+
+    def validate_schedule(self, schedule):
+        _verify_unique_measurement_keys(
+            s.operation for s in schedule.scheduled_operations)
+        for scheduled_operation in schedule.scheduled_operations:
+            self.validate_scheduled_operation(schedule, scheduled_operation)
+
+    def at(self, row: int, col: int) -> Optional[GridQubit]:
```

Comment:
The ion trap is a line instead of a grid, right? Seems you would use `LineQubit` and one parameter instead.